solc/CommandLineInterface.cpp

```diff
@@ -1244,20 +1245,28 @@ void CommandLineInterface::assembleYul(yul::YulStack::Language _language, yul::Y
 				DebugInfoSelection::Default()
 		);
 
-		if (!stack.parseAndAnalyze(src.first, src.second))
-			successful = false;
+		successful = successful && stack.parseAndAnalyze(sourceUnitName, yulSource);
+		if (!successful)
+			solAssert(stack.hasErrors(), "No error reported, but parsing/analysis failed.");
 		else
-			stack.optimize();
-
-		if (successful && m_options.compiler.outputs.asmJson)
 		{
-			std::shared_ptr<yul::Object> result = stack.parserResult();
-			if (result && !result->hasContiguousSourceIndices())
+			if (
+				m_options.compiler.outputs.asmJson &&
+				stack.parserResult() &&
+				!stack.parserResult()->hasContiguousSourceIndices()
```

Comment:
Before this change, `stack.optimize()` was called before invoking `stack.parserResult()`. However, now it is called afterward (on line 1264). If I’m not mistaken, the `Object` returned by `parserResult` could be modified by `optimize()`, meaning that `hasContiguousSourceIndices` would not be checked against the optimized object, no?

Reply:
I don't think that was intentional. The check was inserted after `optimize()` probably simply because it was more convenient to do it there. IMO it makes more sense to do such checks on the unoptimized code that the user is submitting rather than on some modified version so the change is good.

In any case, it should not affect solc behavior in practice - `optimize()` theoretically could change the code in an arbitrary way, but in practice it will never touch the `@use-src` annotation so this property should remain unchanged.

We also have a second validation against this in the asm export code, which would catch this if it did happen: https://github.com/ethereum/solidity/blob/ea52aca0ca816df7c978485205fc08c8a7483f20/libevmasm/Assembly.cpp#L509-L514

We have the check in the CLI only because we don't catch `AssemblyImportError` there. Perhaps we should later change it into a `CodeGenerationError` so that it gets caught by `YulStack` and reported as a normal error.

Reply:
> We have the check in the CLI only because we don't catch `AssemblyImportError` there. Perhaps we should later change it into a `CodeGenerationError` so that it gets caught by `YulStack` and reported as a normal error.

Yeah, I don't see why not do that.

Reply:
ok, I'll keep it in mind for when we're done with this series of PRs and can actually handle those errors properly.